framework for openbmc support, support some rpower and rinv commands

xuweibj opened the pull request:
Supported command:
``rpower <node> status/on/off/reset/boot`` are done based on current urls.
``rinv <node> cpu/dimm/bios/all`` are done based on current urls.

All "print" for debug.

ENV configuration when test:

``export OPENBMC_DEVEL=YES`` in /etc/profile
``source /etc/profile`` load ENV

``restartxcatd``

Node Definition:
```
# lsdef redfishtest
Object name: redfishtest
    bmc=9.3.185.29
    bmcpassword=0penBmc
    bmcusername=root
    groups=test
    mgt=openbmc
    postbootscripts=otherpkgs
    postscripts=syslog,remoteshell,syncfiles,setupntp
```

rinv command:
```
# rinv redfishtest bios
redfishtest: BIOS is_fru : 1
redfishtest: BIOS version :
redfishtest: BIOS fru_type : SYSTEM
redfishtest: BIOS present : True
redfishtest: BIOS fault : False
```

rinv command with XCATBYPASS=1:
```
# XCATBYPASS=1 rinv redfishtest bios
$VAR1 = 'RINV_REQUEST';
$VAR2 = 'RINV_RESPONSE';
$VAR3 = 'LOGIN_REQUEST';
$VAR4 = 'LOGIN_RESPONSE';
$VAR5 = 'LOGIN_RESPONSE';
$VAR6 = 'RINV_REQUEST';

$VAR1 = 'redfishtest';
$VAR2 = {
          'password' => '0penBmc',
          'bmc' => '9.3.185.156',
          'cur_status' => 'LOGIN_REQUEST',
          'username' => 'root'
        };

redfishtest: POST https://9.3.185.156/login -d {"data": [ "root", "0penBmc" ] }
redfishtest: login_response 200 OK
redfishtest: GET https://9.3.185.156/org/openbmc/inventory/enumerate
redfishtest: rinv_response 200 OK
redfishtest: BIOS is_fru : 1
redfishtest: BIOS version :
redfishtest: BIOS fru_type : SYSTEM
redfishtest: BIOS present : True
redfishtest: BIOS fault : False
```

whowutwut commented:
@xuweibj   I will close @gurevichmark pull request https://github.com/xcat2/xcat-core/pull/2569 since you are further along, but I'd like to point out the exercise that we were doing initially,  

We do want to print out debug messages showing which endpoint we are connecting against... so it would be useful to have that set up.  

I'd also like to block the functions from running unless we do something like `export OPENBMC_DEVEL=TRUE`.  Then in our development and UT, we would export this to allow commands to run.  Otherwise it would simply print "This function is not supported at this time"  This allows us to ship this new plugin and have it turned off.  

Can you add those framework in additional commits in this PR?

xuweibj commented:
@chenglch @immarvin @hu-weihua Please help to give some comments, but please do not merge.
Thank you!

whowutwut commented:
@xuweibj  After a review has started, could we avoid using the ``--amend`` option for commit.  This will allow us to save some time by seeing the delta instead of looking at all the changes again.

whowutwut commented:
Can you show the output when doing the rpower command and "XCATBYPASS=TRUE" in the example above in the description?

whowutwut commented on xCAT-server/lib/xcat/plugins/openbmc.pm:
```diff
@@ -0,0 +1,599 @@
+    },
+    RPOWER_STATUS_REQUEST  => {
+        method         => "GET",
+        init_url       => "/org/openbmc/settings/host0",
```
The `/org/openbmc' should be put into a variable so we can easily change this if needed in the future...

whowutwut commented on xCAT-server/lib/xcat/plugins/openbmc.pm:
```diff
@@ -0,0 +1,599 @@
+
+        my $subcommand = $ARGV[0];
+
+# now, only support status, delete when other command supported
```
This section should be indented here..

whowutwut commented on xCAT-server/lib/xcat/plugins/openbmc.pm:
```diff
@@ -0,0 +1,599 @@
+#----------------------------------------------------------------
+
+        if ($subcommand eq "on") {
+            $next_status{LOGIN_RESPONSE} = "RPOWER_ON_REQUEST";
```
I think we should create some functions for this... 
the LOGIN code should be call from within the other functions instead of here on this if/else tests and we should keep track of whether we have already authenticated, and not try to login again... 

so if we had... 
```
login() {
   if logged in, return 
    log in
} 

on() {
   login()
   stuff....
} 
off() { 
    login()
    stuff....
} 
```
Then things like boot... we can do...
```
boot() { 
   login()
   off()
   on() 
} 
```

whowutwut commented on xCAT-server/lib/xcat/plugins/openbmc.pm:
```diff
@@ -0,0 +1,599 @@
+            return ([ 1, "Only 'cpu','dimm','all' are supportted at the same time" ]);
+        }
+    }
+
```
Here's where I would suggest that we figure out all the commands that should be supported here and define those functions as part of the skeleton framework. From there, we create a function that checks .... 
```
unsupported() { 
    if OPENBMC_DEVEL != TRUE
        print "This command is not currently supported" 
        exit 
}
```

Then all the functions will call the unsupported() function... Now we can safely ship this code "turned off" and devel and fvt would export that OPENBMC_DEVEL=TRUE when testing this new code.

whowutwut commented on xCAT-server/lib/xcat/plugins/openbmc.pm:
```diff
@@ -0,0 +1,599 @@
+        }
+    }
+
+    print Dumper(%next_status) . "\n";
```
This is debug code? should be shipped off?  Can we test XCATBYPASS and print it if it's set?  Then we don't need to comment and uncomment if this debug is helpful.

whowutwut commented on xCAT-server/lib/xcat/plugins/openbmc.pm:
```diff
@@ -0,0 +1,599 @@
+
+    if ($next_status{ $node_info{$node}{cur_status} }) {
+        if ($node_info{$node}{cur_status} eq "RPOWER_STATUS_RESPONSE") {
+            if ($response_info->{'data'}->{system_state} =~ /HOST_POWERED_ON/) {
```
@zet809  asked this question on the interlock call. you can also ask them in slack, there should be ENUMs that we might want to define for the status. One suggestion is to parse the yaml file and take the ENUM / description there, may be neat if we did that.. then any firmware implementation we could get the enum/description they want to express from the yaml file... ...

whowutwut replied:
The yaml is part of https://github.com/openbmc/phosphor-dbus-interfaces 
I haven't found a good example to look at.. but just pointing you to this FYI

immarvin commented on xCAT-server/lib/perl/xCAT/OPENBMC.pm:
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/perl
+## IBM(c) 2107 EPL license http://www.eclipse.org/legal/epl-v10.html
+
+package xCAT::OPENBMC;
```
the file name and package name should better be in lower case

whowutwut replied:
@xuweibj  can you respond to this?

xuweibj replied:
@whowutwut , I have talked with yang song about this and decided that don't modify it.

whowutwut commented:
@xuweibj  Thanks for adding the output above! 

I'm wondering what is the value of printing the following :
```
$VAR1 = 'RINV_REQUEST';
$VAR2 = 'RINV_RESPONSE';
$VAR3 = 'LOGIN_REQUEST';
$VAR4 = 'LOGIN_RESPONSE';
$VAR5 = 'LOGIN_RESPONSE';
$VAR6 = 'RINV_REQUEST';
```

Rather, I think we should print the endpoint that we are trying to go against, for example: 
```
Endpoint:  GET /org/openbmc/NetworkManager/Interface/enumerate
Response: 200 OK
```
I think we might be able to tell what we are doing based on the GET/POST/.. etc.. , the URL endpoint... and the response code...

xuweibj commented on xCAT-server/lib/xcat/plugins/openbmc.pm:
```diff
@@ -0,0 +1,546 @@
+    }  
+
+    while (1) { 
+        last unless ($wait_node_num);
```
Yes, it worked.
If one node does not have any next status, $wait_node_num will minus 1. If all nodes have done, $wait_node_num will be '0'.

xuweibj commented:
@whowutwut 

```
$VAR1 = 'RINV_REQUEST';
$VAR2 = 'RINV_RESPONSE';
$VAR3 = 'LOGIN_REQUEST';
$VAR4 = 'LOGIN_RESPONSE';
$VAR5 = 'LOGIN_RESPONSE';
$VAR6 = 'RINV_REQUEST';
```
This is used like status machine.
For this scenario, status machine is ``login request->login response ->rinv request -> rinv response``.

whowutwut commented:
@xuweibj  The above comment, I can't follow how you got `login request->login response ->rinv request -> rinv response`.  

From the messages printed seems like it is... 
`rinv request -> rinv response (??) -> login request -> login response -> login response (??) -> rinv request (should be response?)`.    Seems like we have a few more steps than expected or maybe i'm missing some higher logic...

whowutwut commented:
:+1:  Thanks for this..
```
redfishtest: POST https://9.3.185.156/login -d {"data": [ "root", "0penBmc" ] }
redfishtest: login_response 200 OK
redfishtest: GET https://9.3.185.156/org/openbmc/inventory/enumerate
redfishtest: rinv_response 200 OK```

Can we add some header so that we can easily distinguish normal output and debug output? 

`<node name>: DEBUG POST http....`

Then we can do things like ... `XCATBYPASS=1; rinv redfishtest bios | grep DEBUG`

xuweibj commented:
@whowutwut 

About this hash. For each node, initial status is ``LOGIN_REQUEST``. In this hash table ``$VAR3`` is it, so get its next status is ``$VAR4``. So the status machine is 3->4(5)->6(1)->2.

```
$VAR1 = 'RINV_REQUEST';
$VAR2 = 'RINV_RESPONSE';
$VAR3 = 'LOGIN_REQUEST';
$VAR4 = 'LOGIN_RESPONSE';
$VAR5 = 'LOGIN_RESPONSE';
$VAR6 = 'RINV_REQUEST';
```

whowutwut commented:
@xuweibj  What is the scope for this pull request? It seems like you are adding more and more function in each day.. this is not the purpose here...   I'd like to have a defined scope for this PR and we can merge it in, Is it for framework?  If so, then let's scaffold in the framework which will allow mode developers to co develop functions. 

I'd like to implement the OPENBMC_DEVEL environment variable that will turn on and off this support.  I don't want to merge this code in without this switch.   Then we are able to scaffold the framework and actually have the all `handled_commands` defined... and just "turned off" or blocked.  After they are ready and tested by FVT, we remove the call to the function

To bypass this check during dev and fvt, we would export this variable and it will turn everything on, but this allows us to ship code turned off and protect the users from accidentally running it.

xuweibj commented:
@whowutwut I have added ``OPENBMC_DEVEL`` environment variable and updated description. 
All commands are based on the current investigation. And for ``rinv``, did not filter information now, just show what got. 
Please help to review again. Thank you.

whowutwut commented on xCAT-server/lib/xcat/plugins/openbmc.pm:
```diff
@@ -0,0 +1,604 @@
+
+    $callback  = shift;
+
+#-------------------------------------------------------
```
My idea here is that we would have function that we can call to enable and disable at a more fine control.  
```
sub unsupported { 
    if ($::OPENBMC_DEVEL ne "YES") {
        xCAT::SvrUtils::sendmsg("This function is currently not supported\n", $callback);
        return;
    } else {
        xCAT::SvrUtils::sendmsg("Warning: Currently running development code, use at your own risk\n",  $callback);
    }
}   
```

Then we would call `unsupported()` in all the functions that are planned to be supported in the parse_args..  this is part of the "scaffold framework code" that should have been done first...  @zet809  ...   And ONLY after we have FVT run through the commands testing ... do we open it up by removing the call to `unsupported()` and make a note in the release notes. .  This is how we can roll out support in phases and control it exactly.

whowutwut commented:
If ready to merge in, please change the title...

whowutwut commented:
Also I think @zet809  and @gurevichmark  may not be actively reviewing...  If not, please remove them from the reviewer list so we do not hold up the review/merge of the PR

zet809 commented:
Looks good to me, thx a lot. One potential issue is the "next_status" won't work on Service Node. Pls modify.

whowutwut commented:
Linking PR to Parent Issue #2601
